warehouse/accounts/services.py

```diff
@@ -621,6 +626,79 @@ def get_password_timestamp(self, user_id):
         user = self.get_user(user_id)
         return user.password_date.timestamp() if user.password_date is not None else 0
 
+    def check_device_valid(self, user_id: str, device_id_secret_b64: str) -> bool:
+        """
+        Checks whether the given device id and secret are valid for the given user.
+        """
+        self._metrics.increment("warehouse.authentication.check_device_valid.start")
+
+        if not device_id_secret_b64:
+            return False
+        try:
+            device_id_secret = DeviceIdSecret.from_base64(device_id_secret_b64)
+        except (KeyError, ValueError):
+            return False
+
+        self._check_ratelimits(
+            userid=user_id,
+            tags=["mechanism:check_device_valid"],
+        )
+
+        user = self.get_user(user_id)
+        for device in user.user_devices:
+            if device.device_id != device_id_secret.device_id:
+                continue
+
+            is_expired = (
+                datetime.datetime.now() - device.saved_date
+                > datetime.timedelta(days=SAVED_DEVICE_VALID_DAYS)
+            )
+            if is_expired:
+                continue
+
+            if self.hasher.verify(device_id_secret.secret, device.device_secret):
+                self._metrics.increment(
+                    "warehouse.authentication.check_device_valid.ok"
+                )
+                return True
+
+        self._metrics.increment("warehouse.authentication.check_device_valid.failure")
+        return False
+
+    def generate_device_id_secret(self, user_id: str) -> str:
+        """
+        Generates and saves to the database a new device id and secret for the given
+        user. Returns the device id and secret as a base64 encoded string.
+        """
+        user = self.get_user(user_id)
+
+        device_id = self._generate_device_id(user)
+        device_secret = secrets.token_urlsafe(DEVICE_SECRET_BYTES)
+
+        self.db.add(
+            UserDevice(  # type: ignore
+                user=user,
+                device_id=device_id,
+                device_secret=self.hasher.hash(device_secret),
+            )
+        )
```

Comment:
We should add an [EventTag](https://github.com/pypi/warehouse/blob/main/warehouse/events/tags.py#L60) and [record an event](https://github.com/pypi/warehouse/blob/main/warehouse/accounts/services.py#L568) here

Reply:
I'll add that!

Reply:
I'm noticing that the majority of calls to `record_event` happen in view classes and not in database service classes. Would it be better to add the event to `warehouse/accounts/views.py` when I call `generate_device_id_secret()` (which will be in https://github.com/ristomcgehee/warehouse/pull/2).

Reply:
I've added the recording of the event to my [phase 2 PR](https://github.com/ristomcgehee/warehouse/pull/2). I believe this MR is ready to be merged; I don't think the readthedocs check is failing due to changes introduced in my PR.